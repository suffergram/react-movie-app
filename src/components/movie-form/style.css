.modal-genre-select {
  text-transform: capitalize;
}

.modal-submit-error {
  position: absolute;
  width: 100%;
  margin-top: 95px;
  font-size: 14px;
  text-transform: none;
}

.modal-input-error {
  outline: 1px solid var(--color-1);
}
