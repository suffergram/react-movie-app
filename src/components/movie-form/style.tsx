import styled, { css } from 'styled-components';

export const StyledForm = styled.form`
  width: 856px;
  display: grid;
  grid-template-columns: 2fr 1fr;
  gap: 30px;
`;

export const StyledLabel = styled.label`
  color: var(--color-1);
  font-size: 16px;
  margin: 0 0 13px;
  text-transform: uppercase;
  display: flex;
  flex-direction: column;
  gap: 13px;
`;

export const StyledInput = styled.input<{ $error?: boolean }>`
  background-color: var(--color-form-input);
  border: none;
  padding: 16px;
  font-size: 20px;
  color: var(--color-white);

  ${(props) => {
    if (props.$error) {
      return css`
        outline: 1px solid var(--color-1);
      `;
    }
    return css``;
  }}
`;

export const SubmitErrorMessage = styled.p`
  position: absolute;
  width: 100%;
  margin-top: 95px;
  font-size: 14px;
  text-transform: none;
`;

export const TextareaContainer = styled(StyledLabel)`
  grid-column: auto / span 2;
`;

export const StyledTextarea = styled.textarea`
  background-color: var(--color-form-input);
  border: none;
  padding: 16px;
  font-size: 20px;
  color: var(--color-white);
  box-sizing: border-box;
  resize: none;
  width: 100%;
  height: 197px;
`;

export const ButtonContainer = styled.div`
  grid-column: auto / span 2;
  display: flex;
  gap: 13px;
  justify-self: end;
  margin-top: 30px;
`;
