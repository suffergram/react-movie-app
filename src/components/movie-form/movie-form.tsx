import { useForm, SubmitHandler, Controller } from 'react-hook-form';
import Button from '../button/button';
import { FormInput } from '../../types/form-input';
import { FormField } from '../../types/edit-form-field';
import GenreSelect from '../genre-select/genre-select';
import './style.css';

type FormProps = {
  onSubmit: SubmitHandler<FormInput>;
  params?: FormInput | undefined;
};

const initialParams: FormInput = {
  title: '',
  release_date: '',
  poster_path: '',
  vote_average: 0,
  genres: [],
  runtime: 0,
  overview: '',
};

export default function MovieForm({
  onSubmit,
  params = initialParams,
}: FormProps) {
  const {
    register,
    handleSubmit,
    formState: { errors },
    control,
    reset,
  } = useForm<FormInput>({
    defaultValues: params,
  });

  const requiredMessage = 'This is required';

  const handleReset = () => reset();

  return (
    <form className="modal-edit-form" onSubmit={handleSubmit(onSubmit)}>
      <label className="movie-form-label">
        title
        <input
          className={
            errors[FormField.Title]?.message ? 'modal-input-error' : undefined
          }
          placeholder="Select Title"
          inputMode="text"
          {...register(FormField.Title, { required: requiredMessage })}
        />
        <p className="modal-submit-error">{errors[FormField.Title]?.message}</p>
      </label>
      <label className="movie-form-label">
        release date
        <input
          className={
            errors[FormField.ReleaseDate]?.message
              ? 'modal-input-error'
              : undefined
          }
          type="date"
          placeholder="Select Date"
          {...register(FormField.ReleaseDate, {
            required: requiredMessage,
          })}
        />
        <p className="modal-submit-error">
          {errors[FormField.ReleaseDate]?.message}
        </p>
      </label>
      <label className="genre-select-container movie-form-label">
        genres
        <Controller
          control={control}
          name={FormField.Genres}
          rules={{
            validate: (value) => (value.length === 0 ? requiredMessage : true),
          }}
          render={({ field: { value, onChange }, fieldState: { error } }) => (
            <GenreSelect
              value={value}
              onChange={onChange}
              className={error ? 'modal-input-error' : undefined}
            />
          )}
        />
        <p className="modal-submit-error">
          {errors[FormField.Genres]?.message}
        </p>
      </label>
      <label className="movie-form-label">
        runtime
        <input
          className={
            errors[FormField.Runtime]?.message ? 'modal-input-error' : undefined
          }
          placeholder="minutes"
          inputMode="numeric"
          type="number"
          step="1"
          min="0"
          max="999"
          {...register(FormField.Runtime, { required: requiredMessage })}
        />
        <p className="modal-submit-error">
          {errors[FormField.Runtime]?.message}
        </p>
      </label>
      <label className="movie-form-label">
        movie url
        <input
          className={
            errors[FormField.PosterPath]?.message
              ? 'modal-input-error'
              : undefined
          }
          placeholder="https://"
          inputMode="url"
          {...register(FormField.PosterPath, {
            required: false,
            pattern: {
              value: /((https)|(HTTPS)):\/\/\w+\.\w+/,
              message: 'Invalid path',
            },
          })}
        />
        <p className="modal-submit-error">
          {errors[FormField.PosterPath]?.message}
        </p>
      </label>
      <label className="movie-form-label">
        rating
        <input
          placeholder="7.8"
          inputMode="numeric"
          type="number"
          lang="en"
          step="0.1"
          min="0"
          max="10"
          {...register(FormField.VoteAverage, {
            required: false,
            min: 0,
            max: 10,
          })}
        />
      </label>
      <label className="movie-form-label">
        overview
        <textarea
          placeholder="Movie description"
          {...register(FormField.Overview, { required: false })}
        />
      </label>
      <div>
        <Button className="cancel modal-edit-button" onClick={handleReset}>
          reset
        </Button>
        <Button className="confirm modal-edit-button" type="submit">
          submit
        </Button>
      </div>
    </form>
  );
}
