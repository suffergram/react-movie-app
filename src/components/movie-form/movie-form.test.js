import { render, screen } from "@testing-library/react";
import { MovieForm } from "./movie-form";

describe('Movie form', () => {
  it('Matches the snapshot', () => {
    render(
      <MovieForm />
    );

    const form = screen.getByLabelText('movie-form');

    expect(form).toMatchSnapshot();
  });
});