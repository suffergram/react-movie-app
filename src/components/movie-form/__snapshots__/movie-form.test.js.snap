// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie form Matches the snapshot 1`] = `
<form
  aria-label="movie-form"
  class="modal-edit-form"
>
  <label
    class="movie-form-label"
  >
    title
    <input
      inputmode="text"
      name="title"
      placeholder="Select Title"
    />
    <p
      class="modal-submit-error"
    />
  </label>
  <label
    class="movie-form-label"
  >
    release date
    <input
      name="release_date"
      placeholder="Select Date"
      type="date"
    />
    <p
      class="modal-submit-error"
    />
  </label>
  <label
    class="genre-select-container movie-form-label"
  >
    genres
    <input
      aria-expanded="false"
      aria-haspopup="dialog"
      class="genre-select"
      placeholder="Select Genre"
      readonly=""
      value=""
    />
    <p
      class="modal-submit-error"
    />
  </label>
  <label
    class="movie-form-label"
  >
    runtime
    <input
      inputmode="numeric"
      max="999"
      min="0"
      name="runtime"
      placeholder="minutes"
      step="1"
      type="number"
    />
    <p
      class="modal-submit-error"
    />
  </label>
  <label
    class="movie-form-label"
  >
    movie url
    <input
      inputmode="url"
      name="poster_path"
      placeholder="https://"
    />
    <p
      class="modal-submit-error"
    />
  </label>
  <label
    class="movie-form-label"
  >
    rating
    <input
      inputmode="numeric"
      lang="en"
      max="10"
      min="0"
      name="vote_average"
      placeholder="7.8"
      step="0.1"
      type="number"
    />
  </label>
  <label
    class="movie-form-label"
  >
    overview
    <textarea
      name="overview"
      placeholder="Movie description"
    />
  </label>
  <div>
    <button
      class="cancel modal-edit-button"
      type="button"
    >
      reset
    </button>
    <button
      class="confirm modal-edit-button"
      type="submit"
    >
      submit
    </button>
  </div>
</form>
`;
